server/sonar-ce/src/main/java/org/sonar/ce/notification/ReportAnalysisFailureNotificationExecutionListener.java

```diff
@@ -0,0 +1,107 @@
+  private final DbClient dbClient;
+  private final ReportAnalysisFailureNotificationSerializer taskFailureNotificationSerializer;
+  private final System2 system2;
+
+  public ReportAnalysisFailureNotificationExecutionListener(NotificationService notificationService, DbClient dbClient,
+    ReportAnalysisFailureNotificationSerializer taskFailureNotificationSerializer, System2 system2) {
+    this.notificationService = notificationService;
+    this.dbClient = dbClient;
+    this.taskFailureNotificationSerializer = taskFailureNotificationSerializer;
+    this.system2 = system2;
+  }
+
+  @Override
+  public void onStart(CeTask ceTask) {
+    // nothing to do
+  }
+
+  @Override
+  public void onEnd(CeTask ceTask, CeActivityDto.Status status, @Nullable CeTaskResult taskResult, @Nullable Throwable error) {
+    if (status == CeActivityDto.Status.SUCCESS) {
+      return;
+    }
+    String projectUuid = ceTask.getComponentUuid();
+    if (!CeTaskTypes.REPORT.equals(ceTask.getType()) || projectUuid == null) {
+      return;
+    }
+
+    if (notificationService.hasProjectSubscribersForTypes(projectUuid, singleton(ReportAnalysisFailureNotification.TYPE))) {
+      try (DbSession dbSession = dbClient.openSession(false)) {
+        ComponentDto projectDto = dbClient.componentDao().selectOrFailByUuid(dbSession, projectUuid);
+        checkScopeAndQualifier(projectDto);
+        CeActivityDto ceActivityDto = dbClient.ceActivityDao().selectByUuid(dbSession, ceTask.getUuid())
+          .orElseThrow(() -> new RowNotFoundException(format("CeActivity with uuid '%s' not found", ceTask.getUuid())));
+        ReportAnalysisFailureNotification taskFailureNotification = buildNotification(ceActivityDto, projectDto, error);
+        notificationService.deliver(taskFailureNotificationSerializer.toNotification(taskFailureNotification));
+      }
+    }
+  }
+
+  /**
+   * @throws IllegalArgumentException if specified {@link ComponentDto} is not a project.
+   */
+  private static void checkScopeAndQualifier(ComponentDto projectDto) {
+    String scope = projectDto.scope();
+    String qualifier = projectDto.qualifier();
+    checkArgument(
+      scope.equals(Scopes.PROJECT) && qualifier.equals(Qualifiers.PROJECT),
+      "Component %s must be a project (scope=%s, qualifier=%s)", projectDto.uuid(), scope, qualifier);
+  }
+
+  private ReportAnalysisFailureNotification buildNotification(CeActivityDto ceActivityDto, ComponentDto projectDto, @Nullable Throwable error) {
+    Long executedAt = ceActivityDto.getExecutedAt();
+    return new ReportAnalysisFailureNotification(
+      new ReportAnalysisFailureNotification.Project(
+        projectDto.uuid(),
+        projectDto.getKey(),
+        projectDto.name(),
+        projectDto.getBranch()),
+      new ReportAnalysisFailureNotification.Task(
+        ceActivityDto.getUuid(),
+        ceActivityDto.getSubmittedAt(),
+        executedAt == null ? system2.now() : executedAt),
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "now"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)